### PR TITLE
Add support for docker

### PR DESCRIPTION
**Pull request by mbrooks**
This commit adds basic support for docker image creation. A Dockerfile
has been added along with a Makefile that contains targets to assist
with the docker build process.

Run `make docker-build` to build the docker image. After it is built
the image can be run by executing `docker run journalbeat`. Run
`make docker-tag` to build the image and tag the docker image with the
current git tag.

### CODE REVIEW

**Comment by mbrooks** [comment]
Please provide any feedback/direction for this to let me know if I am heading in the right direction. This may be a work in progress for a bit, as I just realized that there is no journalbeat.yml config file made available to this docker image. 

Still Needs to be done:

- [x] journalbeat.yml for docker

**Comment by eghobo** [comment]
@mbrooks: have you tested this image with latest stable CoreOS 1235.9.0?

**Comment by mbrooks** [comment]
@eghobo I have not, but first this is not an image, it is tools to create the image. Secondly, it's docker, so in theory it should work as long as it's running on a linux host machine that supports systemd. 

Is it possible that you could test it out on CoreOS?

**Comment by nrvnrvn** [comment]
@mbrooks why add `docker/journalbeat.yml` when you already have it in `etc`?

**Comment by mbrooks** [comment]
@nicorevin The defaults in `/etc/journalbeat.yml` don't work well in a docker environment. `/docker/journalbeat.yml` is configured to work a little better out of the box. Although, I will admit it is a bit confusing to have two `jouralbeat.yml` files in a project. Perhaps it makes sense to rename it? What do you think?

**Comment by nrvnrvn** [comment]
@mbrooks `/etc/journalbeat.yml` is a documented template which everyone is supposed to amend according to their own needs.

What bothers me most in regard to a Dockerfile is this: https://github.com/mheese/journalbeat/issues/35#issuecomment-274938416

It looks like `systemd` is really keeping its promise about journald interface. But it would be good to prove it. If it is actually true then all the distro-specific builds we are using right now are redundant. And we could just create a generic image containing `journalbeat` binary and `systemd-lib` and put it to `hub.docker.com` along with instructions on how to pass a configuration file and mount all the necessary directories (see https://github.com/mheese/journalbeat/issues/35#issuecomment-256726376).

**Comment by mbrooks** [comment]
@nicorevin Keep in mind, my goal for now is to provide basic support for Docker sponsored by the official project. There are a lot of other docker images out there for journalbeat, but none are official, some don't share the source, most don't keep tagged versions of the releases, and all have a hard time keeping up with the project.  

As for docker, it is standard practice to define ENV variables for the various options you wish to support. Editing a yml config file so that it works for everyone is not something you typically do with docker. That being said, nothing is stopping you from overriding the shipped `journalbeat.yml` with a shared volume if the shipped options do not work for you. But, this is just a configuration issue.

Now, if this project wishes to support other features other than specifying a logstash server, we can do that, but it will probably mean that the `journalbeat.yml` config fill will need to be dynamically generated. As I stated previously, my goal isn't to implement everything someone would want, but to start with some sort of basic support for docker. We can learn what's important to people and implement those features as needed.

As for [#35 comment](https://github.com/mheese/journalbeat/issues/35#issuecomment-274938416), docker should have pretty good support for any systemd based distro. This docker image, however, will never be able to full support all distros and platforms as that gets really complicated. For the docker image to work as expected, the host kernel must be compatible enough to run debian jessie and the officially supported golang docker image. Docker is just not designed to be 100% compatible for every situation but it typically comes down to if the kernels are compatible. Most projects support x86 first and the standard distros. It is possible to support other platforms such as ARM, m68k, etc within this project by using tagged versions such as `5.1.1-arm`, however, this is not the problem I am trying to solve today in this PR.

**Comment by mbrooks** [comment]
BTW, It's perfectly acceptable to reject this PR and say that Docker doesn't need to live with this project. That maybe a better solution. It really depends on if everything should be kept together, or separated. If this is rejected, however, I would really want to work closely with you guys toward creating an official journalbeats docker image and make sure we do this correctly like having the project under the official repository, etc.

**Comment by nrvnrvn** [comment]
I will try to clarify my point.

```
FROM ubuntu:16.04

COPY journalbeat .

ENTRYPOINT ["/journalbeat"]
CMD ["--help"]
```

and running
```sh
$ docker build -t journalbeat .
$ docker run --rm \
-v /tmp/etc/journalbeat.yml:/journalbeat.yml \
-v /etc/machine-id:/etc/machine-id:ro \
-v /var/log/journal:/var/log/journal:ro \
-v /run/systemd/journal:/run/systemd/journal:ro \
-v /run/log:/run/log:ro \
journalbeat -N -e -d '*' -c journalbeat.yml
```
works for me and does everything needed.
( @eghobo ⬆️ has been tested in CoreOS stable ;) )

`journalbeat` binary in this example is pre-built in debian jessie and simply copied on build to the new container. 
`ubuntu`, `debian`, `fedora`, `centos` images include everything needed to run journalbeat - namely libsystemd:

```
$ docker run --rm -ti ubuntu:16.04 dpkg -L libsystemd0 | grep ".so"
/lib/x86_64-linux-gnu/libsystemd.so.0.14.0
/lib/x86_64-linux-gnu/libsystemd.so.0
```

`ubuntu` was taken because of the later version of systemd but that does not matter much. It is just me who believes that more recent versions of software are better than the earlier ones. :)

The idea is just to wrap the journalbeat binary in a container and give it (mount via `-v`) all the bits necessary, i.e. actual log directories, config file and cursor and pending queue files.

Have a look at https://hub.docker.com/_/logstash/. The same idea is implemented there.
Having an image built using the above Dockerfile a person will have two options:
- do smth like the above command line example - just use the image and mount everything including the config and pass all the command line options directly;
- use the `journalbeat` image to make smth like:
```
FROM journalbeat:5.2.1

COPY journalbeat.yml .

CMD ["-e", "-c", "journalbeat.yml"] # or whatever is needed
```

Thus we will have:
- a basic image containing journalbeat;
- an instruction on how to use it or extend it.

**Comment by mbrooks** [comment]
@nicorevin That is definitely another approach to this and I think I'm okay with that. 

I do want to keep the Makefile in the project so that we can automate the build process and make it easy to run in CI or from the command line. Also, I want to add an option for BUILD_IN_DOCKER that will build the journalbeat in a docker container so that it can be built in MacOS or CI. 

Does that sound good?

**Comment by nrvnrvn** [comment]
That sounds good.

Ideally it would be good to bind docker binary release and docker image creation and publishing.
See: 
https://github.com/mheese/journalbeat/releases
https://github.com/mheese/journalbeat/blob/master/.travis.yml
https://sebest.github.io/post/using-travis-ci-to-build-docker-images/

What do you think?

**Comment by mbrooks** [comment]
@nicorevin I would like to hookup travis eventually, but I am not a fan of using CI to script out the build. I prefer to use make or a bash script so that the builds can be run and tested on any local machine. CI should be able to execute any script once it's made.

**Comment by eghobo** [comment]
@mbrooks: i tested this PR with CoreOS stable (1235.9.0) and SUSE Linux Enterprise Server 12 SP2 and it works. I agree with @nicorevin that it's better to use binary then build it inside container. But journalbeat container should have default config and user can overbite this config if need it (we do it within Kub all the time).

**Comment by eghobo** [comment]
just to test the idea ;) https://gist.github.com/eghobo/eb2396567b6b1bcf9ef15dfa5bc4d2c4 - works at CoreOS and SLES as well.

**Comment by mheese** [comment]
okay, it looks like I need to chime in on the whole docker discussion a bit more.

First of all, thanks @mbrooks for putting this together! Much appreciated!

You've mentioned above that there are a lot of unofficial docker images out there, can you point me to all the different unofficial docker images that you have seen/used so far? I kind of want to get a sense for what people are doing and trying to achieve.

It also looks to me that this docker image fullfills two purposes: compilation and running journalbeat. While this is no problem at all, I personally always try to keep my images to a bare minimum.

That said, I have a feeling we should rather merge this early on and make updates as necessary. It feels like we are blocking a lot of people from using it.

In terms of support, I think it is most important to get this working on CoreOS - as this is the only way how you can properly integrate with it on that platform. That said, I'd like to see support for the following OSes:
- CoreOS
- CentOS / RHEL
- Debian / Ubuntu
- SLES

Long-term goal should also be that this can be used as log shipper for a Kubernetes cluster.

I actually only like to see one change if you want to: you can mention yourself as maintainer there in the Dockerfile!

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM golang:1.8 
+MAINTAINER mheese
```
Comment (mheese) [inline docker/Dockerfile]: @mbrooks you can mention yourself here!

**Comment by mbrooks** [comment]
@mheese The two docker images that I have used are:
https://hub.docker.com/r/sebjay/journalbeat/
https://hub.docker.com/r/mcasimir/journalbeat/
Note: Neither of the above have links back to their sources in github, which isn't typical.

Full list of docker images on Dockerhub:
https://hub.docker.com/search/?isAutomated=0&isOfficial=0&page=1&pullCount=0&q=journalbeat&starCount=0

Also, I think I agree with the minimum approach and removing the build from the Dockerfile, but I have to do it in a way that's compatible with CI. I'm indifferent on if we should provide a default config, but I'm leaning towards yes we should.

**Comment by mheese** [comment]
@mbrooks I think for now we can go with the build _within_ the `Dockerfile` approach, however, please think about how we could improve that because I also think that building it with docker is actually the right approach. I guess we need just 2 `Dockerfile`s

**Comment by mbrooks** [comment]
@mheese I believe you are correct, and I was actually planning on modifying what I've done to do that. I was thinking we would have 2 docker files, once the build has completed, you can copy the image from the docker image into the build path, then run the second docker file to build the image we would ship out. I may try to knock this out tonight. I should have some free time.

**Comment by mbrooks** [comment]
@mheese We now have two docker files. Running `make docker-build` now builds the `journalbeat` binary by coping the source to a container and running `dockerfile.build`. After the build is successful, the binary is then copied from the image into `./build` along with `dockerfile.release`. This then builds a slimmer more streamlined release without the source code in the container.

@nicorevin I tried to update the instructions to show how to include the docker base image and override the default options, but without knowing the final name of this repo and tags and such, it's a bit confusing right now. It may make more sense once we have a release in dockerhub to circle back around and fix the instructions.

**Comment by kainz** [comment]
So in the vein of the above discussion, at $WORK I also use journalbeat in our mesos + consul + marathon pipeline.

For that environment, I build an image similar to the way in which the above people have and bake it into my OS images.  The interesting part is that I use consul-template to dynamically update my logstash targets as necessary.

Given all the variants of how one would compile, configure, and execute journalbeat, maybe a contrib/$ENVIRONMENT folder is in order?

As one example, here's my consul-template template for deriving consul-discovered logstash servers:
```
{{/* begin lookup prep*/}}
{{ $requiredTags :=  "prod,beats" | split "," | toJSON | parseJSON }}
{{ $lookup := service "TAGNAME.LOGSTASH-SERVICE-NAME-HERE|any" }}
{{/* This oddity is here to ensure that the scratchMap "targets" gets created *with the correct type*. Thanks, HashiCorp. */}}
{{ define "first" }}{{range $i, $v := .}}{{if eq $i 0}}{{$v.Node}}:{{$v.Port}}{{end}}{{end}}{{end}}
{{ scratch.MapSet "targets" (executeTemplate "first" $lookup) "unset" }}
{{ range $s := $lookup}}{{ if containsAll $requiredTags .Tags}}{{scratch.MapSet "targets" (print $s.Node ":" $s.Port) (print $s.Address ":" $s.Port) }}{{end}}{{end}}
{{/* end lookup */}}
journalbeat:
  seek_position: cursor
  cursor_seek_fallback: head
  write_cursor_state: true
  cursor_state_file: /journalbeat/journalbeat-cursor-state
  clean_field_names: true
name: journalbeat
processors:
    - add_cloud_metadata:
output.elasticsearch:
  enabled: false
output.logstash:
    max_retries: 1
    enabled: true
    hosts: {{ scratch.MapValues "targets" | toJSON }}
    worker: 1
    compression_level: 3
    timeout: 2
logging.metrics.enabled: true
logging.to_files: false
```

**Comment by mbrooks** [comment]
@kainz That looks like a j2 template. Unfortunately, I'm not familiar enough with how consul manages containers to give an answer for this. But, if you can pass ENV vars with consul instead of building an image and only need to define one logstash server, then this docker container should work for you out of the box.

**Comment by kainz** [comment]
@mbrooks It's a Go template.  Superficially similar to j2, but just different enough to make you throw a laptop through the window when you need to do anything sufficiently interesting.

The trick isn't assigning a server at startup time, its dealing with your serverlist changing over time, which will be true for pretty much any service discovery environment.  A separate thing to deal with, the generic way to interface to that seems to be to add SRV DNS lookup support to infrastructure software, but since that's by no means a global tool to use, things like consul-template or containerpilot become the order of the day when you want to dynamically reconfigure a running app to be aware when your server list for some service changes.

I'm talking about the concern of service discovery, which is fairly separate from container management, except when the lifecycle of a process is concerned.

**Comment by mbrooks** [comment]
@kainz I'm still trying to wrap my head around what you're saying. Are you saying you use that template for your `journalbeat.yml`? Or do you feed it through some other application?

If you're looking for a way to replace `journalbeat.yml` with your own version during build, then you can simply copy your custom `journalbeat.yml` to `build/journalbeat.yml` before running `make docker-build` and it will detect and respect your custom `journalbeat.yml` during the build and use it.

**Comment by mheese** [comment]
I think that this is a very good start. It divides the concerns of building and packaging, and we probably should rethink the current Travis integration in the light of this.

It also does exactly what @nicorevin mentioned which is very important: it provides a good _base_ docker image which people can easily extend on:
- run it as a docker container as mentioned in the instructions, adjusting the LOGSTASH_HOST is all you need ... this is perfect for a quickstart with journalbeat and docker
- you can easily mount in your own journalbeat configuration file, and you don't really need to change anything for it ( @mbrooks , maybe update that docker readme with this idea as well ) 
- you can extend from this docker image and build your own: either integrate your own config file, do fancy service discovery as @kainz mentioned

For official releasing, we could use my `mheese` docker hub namespace? I like when github and dockerhub names match ... a `docker pull mheese/journalbeat` would be all one needs.

We also should think further for other integrations:
1. as @kainz mentioned, he is using it to integrate mesos + consul + marathon ... it would be awesome if we can have a dockerfile and build for this as well ...
2. as I mentioned myself in this thread: I'd like to see an image that we can plug in as a kubernetes log shipper, that people could run instead of the default fluentd daemon sets.

I also don't believe that we need a separate contrib folder for this, this can all live within the docker folder IMHO.

We can always work with docker tags to get the right images ... like a lot of official images (https://hub.docker.com/_/node/ e.g.) they are working with `VERSION-alpine`, etc. to tag particular versions, we could do the same here: `VERSION-mesos`, `VERSION-kubernetes` ... and if there are really problems occuring in the future, we can make `VERSION-ubuntu`, `VERSION-centos`, etc. which are using ubuntu or centos instead of debian:jessie as the base image.

@mbrooks would you be willing to take a lead on the direction of this? If yes, please create a new / or several new issues and reference or take over the discussion from this PR. The thread also becomes a bit too long for a discussion about a simple PR.

Apart from all this, LGTM! :shipit:

**Review by mheese — APPROVED** [review]

**Comment by mbrooks** [comment]
@mheese I'll create issues for the suggested when I have time.

**Pull request merged**